File: test/reference/shake_hashlib.py
"""Compare CTL SHAKE128/256 with Python's hashlib.

The companion executable prints deterministic answers across the absorb and
squeeze rate boundaries. hashlib is an independent review oracle and is not a
runtime dependency of CTL.
"""

from __future__ import annotations

import hashlib
import subprocess
import sys


def patterned(size: int) -> bytes:
    return bytes((index * 73 + 19) & 0xFF for index in range(size))


def main() -> int:
    if len(sys.argv) != 2:
        print(f"usage: {sys.argv[0]} PATH-TO-SHAKE-PROBE", file=sys.stderr)
        return 2

    process = subprocess.Popen(
        [sys.argv[1]], stdout=subprocess.PIPE, text=True, encoding="ascii"
    )
    assert process.stdout is not None

    checked = 0
    for line in process.stdout:
        family, message_text, output_text, actual = line.rstrip("\n").split("\t")
        message_size = int(message_text)
        output_size = int(output_text)
        message = patterned(message_size)
        constructor = hashlib.shake_128 if family == "1" else hashlib.shake_256
        expected = constructor(message).hexdigest(output_size)
        if actual != expected:
            process.kill()
            print(
                f"SHAKE{family} differs at {message_size}-byte input and "
                f"{output_size}-byte output",
                file=sys.stderr,
            )
            return 1
        checked += 1

    if process.wait() != 0:
        return 1

    expected_count = 2 * 25 * 25
    if checked != expected_count:
        print(f"expected {expected_count} answers, received {checked}", file=sys.stderr)
        return 1

    print(
        f"{checked} SHAKE answers (both rates and adjacent boundaries) "
        "agree with hashlib"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: test/reference/shake_hashlib.cpp
/**
 * @file shake_hashlib.cpp
 * @brief Emits SHAKE answers for the independent hashlib checker
 *
 * Input and output sizes cross both SHAKE rates and their immediate
 * boundaries. Build and compare from the repository root with:
 *
 *   c++ -std=c++17 -O2 -Iinclude test/reference/shake_hashlib.cpp \
 *       -o /tmp/shake_hashlib
 *   python3 test/reference/shake_hashlib.py /tmp/shake_hashlib
 */

#include <cstddef>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <vector>

#include <ctl/hash/shake>

namespace {

const size_t sizes[] = {0,   1,   2,   7,   31,  63,  127, 135, 136,
                        137, 167, 168, 169, 199, 200, 255, 256, 271,
                        272, 273, 335, 336, 337, 511, 512};

std::vector<uint8_t> patterned(size_t size) {
  std::vector<uint8_t> result(size);
  for (size_t i = 0; i < result.size(); ++i)
    result[i] = static_cast<uint8_t>(i * 73u + 19u);
  return result;
}

template <class Xof>
void emit(char family, size_t message_size, ctl::bytes message,
          size_t output_size) {
  std::vector<uint8_t> output(output_size);
  Xof::expand(message, output);

  std::ostringstream encoded;
  encoded << std::hex << std::setfill('0');
  for (uint8_t byte : output)
    encoded << std::setw(2) << static_cast<unsigned int>(byte);
  std::cout << family << '\t' << message_size << '\t' << output_size << '\t'
            << encoded.str() << '\n';
}

template <class Xof> void emit_family(char family) {
  for (size_t message_size : sizes) {
    const std::vector<uint8_t> message = patterned(message_size);
    for (size_t output_size : sizes)
      emit<Xof>(family, message_size, message, output_size);
  }
}

} // namespace

int main() {
  emit_family<ctl::hash::shake128>('1');
  emit_family<ctl::hash::shake256>('2');
}
